cmd/gitlab_server_periodic.go

```diff
@@ -0,0 +1,104 @@
+package cmd
+
+import (
+	"context"
+	"log/slog"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/jippi/scm-engine/pkg/config"
+	"github.com/jippi/scm-engine/pkg/scm"
+	"github.com/jippi/scm-engine/pkg/state"
+	slogctx "github.com/veqryn/slog-context"
+)
+
+func startPeriodicEvaluation(ctx context.Context, interval time.Duration, filter scm.MergeRequestListFilters, wg *sync.WaitGroup) {
+	// Empty interval means disabling
+	if interval == 0 {
+		slogctx.Warn(ctx, "scm-engine will not be doing periodic evaluation since interval is '0'. Set 'SCM_ENGINE_PERIODIC_EVALUATION_INTERVAL' or '--periodic-evaluation-interval'  to a non-zero duration to activate")
+
+		return
+	}
+
+	wg.Add(1) // +1: Periodic Evaluation
+
+	// I can't think of a good reason why anyone would want to have it running more frequently than every 15m, so enforcing a floor value.
+	//
+	// If you are reading this code and need more frequent periodic evaluations then please open a PR or issue with use-case
+	// and I will happily re-evaluate this logic.
+	if interval < 15*time.Minute {
+		slogctx.Warn(ctx, "'SCM_ENGINE_PERIODIC_EVALUATION_INTERVAL' / '--periodic-evaluation-interval' is set to a value less than '15 minutes'; changing the value to '15m'")
+
+		interval = 15 * time.Minute
+	}
+
+	// Initialize the SCM-Engine client
+	client, err := getClient(ctx)
+	if err != nil {
+		panic(err)
+	}
+
+	// Configure logger and custom fields
+	ctx = slogctx.With(ctx,
+		slog.String("subsystem", "periodic_evaluation"),
+		slog.Duration("periodic_evaluation_interval", interval),
+		slog.Any("periodic_evaluation_filters", filter.AsGraphqlVariables()),
+	)
+
+	go func(wg *sync.WaitGroup) {
+		defer wg.Done() // -1: Periodic Evaluation
+
+		ticker := time.NewTicker(interval)
+		defer ticker.Stop()
+
+		slogctx.Info(ctx, "Waiting for first evaluation cycle tick to happen")
+
+		for {
+			select {
+			case <-ctx.Done():
+				slogctx.Info(ctx, "Stopping periodic evaluation as scm-engine is shutting down")
+
+				return
+
+			case <-ticker.C:
+				slogctx.Info(ctx, "Starting periodic evaluation cycle")
+
+				results, err := client.FindMergeRequestsForPeriodicEvaluation(ctx, filter)
+				if err != nil {
+					slogctx.Error(ctx, "Failed to generate merge request list to evaluate", slog.Any("error", err))
+
+					continue
+				}
+
+				for _, mergeRequest := range results {
+					ctx := ctx // make sure we define a
```

Comment:
... fresh context so we don't append to the same context over and over (find link to article about this problem)